### PR TITLE
ScalafmtConfig: define a new `indent.main` parameter

### LINKED ISSUE #1493: Definition indentation tab width doesn't work
Version: 2.0.0-RC4
Integration: VS Code with Scala (Metals) extension
Configuration:
```
  version = "2.0.0-RC4"
  maxColumn = 80
  continuationIndent.callSite = 8
  continuationIndent.defnSite = 8
  continuationIndent.extendSite = 8
```

## Steps

When I run VS code format document (that calls scalafmt), scalafmt formats code like this:
```scala
final class Goods(
        _food: Int = 0, 
        _materials: Int = 0, 
        _products: Int = 0, 
        _fuel: Int = 0 
) {
  var raw = Array(_food, _materials, _products, _fuel)
  def food: Int = raw(0)
  def materials: Int = raw(1)
  def products: Int = raw(2)
  def fuel: Int = raw(3)
  def weight = raw.sum
}
```
## Expectation

I would like the formatted output to look like this:
```scala
final class Goods(
        _food: Int = 0, 
        _materials: Int = 0, 
        _products: Int = 0, 
        _fuel: Int = 0 
) {
        var raw = Array(_food, _materials, _products, _fuel)
        def food: Int = raw(0)
        def materials: Int = raw(1)
        def products: Int = raw(2)
        def fuel: Int = raw(3)
}

### LINKED ISSUE #2385: Please support a 4-space indent.
## Problem

Why does this formatter not support 4-space indentation?
I know that 2-space is widely known in the scala community, as talked about in a past issue ( #1493 ), but I think there are many engineers who want to use 4-space.
Are there any technical difficulties?

## Expectation

support a 4-space indent.

### PR DESCRIPTION
**Pull request by kitbellew**
Fixes #2385. Fixes #1493.

### CODE REVIEW

File: scalafmt-tests/src/test/resources/unit/Import.source [hunk 1]
```diff
@@ -181,17 +185,17 @@ object blah {
 >>>
 object blah {
   import a.{
-    // format: off
+      // format: off
```
Comment (kitbellew) [inline scalafmt-tests/src/test/resources/unit/Import.source:188]: NB: this is a bug that #2397 fixes.

**Review by olafurpg — COMMENTED** [review]
I’m on the fence about this change. Are there feature requests from users asking about these particular settings?

The type signature indentation could maybe share the setting with “defnSite” 🤔 

With Scala 3 and significant indentation around the corner I think we should be careful not to introduce too many new indentation settings.

**Comment by kitbellew** [comment]
@olafurpg not really. i was mainly playing around with the primary indent, partly because of the impending significant indentation and partly because of the general howl around availability of this setting, and wanted to debug some changes one type at a time (and found a couple of bugs along the way, fixing in other changes).

regarding the primary indent: as long as code is not more complex, i don't have any problem with giving people the rope if that's what they want. the perennial tabs-vs-spaces debacle is one example of wisdom of reserving one's opinions.

**Comment by olafurpg** [comment]
My personal opinion remains that Scalafmt shouldn’t support configurable indentation. @kitbellew if your opinion is that this setting should exist then that’s a good enough argument for me. This is not a hill I will die defending! I trust you to make a decision.

**Comment by kitbellew** [comment]
> My personal opinion remains that Scalafmt shouldn’t support configurable indentation. @kitbellew if your opinion is that this setting should exist then that’s a good enough argument for me. This is not a hill I will die defending! I trust you to make a decision.

i think that most people will stick with 2 but usually the most vocal ones are the detractors. as i said, i don't want to tell people how much salt to use in their food. @tgodzik, what do you think?

**Comment by tgodzik** [comment]
> > My personal opinion remains that Scalafmt shouldn’t support configurable indentation. @kitbellew if your opinion is that this setting should exist then that’s a good enough argument for me. This is not a hill I will die defending! I trust you to make a decision.
> 
> i think that most people will stick with 2 but usually the most vocal ones are the detractors. as i said, i don't want to tell people how much salt to use in their food. @tgodzik, what do you think?

I think configurable indentation is fine, especially when it comes to Scala 3, however, I would not split it into too many options. I would prefer to have single option for all of those if possible. Too many options result in verbose configuration and more complexity for maintainers. We've been discussing this back and forth and the overall consensus is that people might think they need all the options, but that just results in useless arguments about the perfect style. Setting one option for indentation though should not cause that mich problem I think. 

That's at least my opinion, I am by far not an expert. I usually am happy with whatever the formatter gives me and each day I am thankful that I don't get formatting comments on most PRs expect in some edge cases.

**Comment by kitbellew** [comment]
> > > My personal opinion remains that Scalafmt shouldn’t support configurable indentation. @kitbellew if your opinion is that this setting should exist then that’s a good enough argument for me. This is not a hill I will die defending! I trust you to make a decision.
> > 
> > 
> > i think that most people will stick with 2 but usually the most vocal ones are the detractors. as i said, i don't want to tell people how much salt to use in their food. @tgodzik, what do you think?
> 
> I think configurable indentation is fine, ... I would prefer to have single option for all of those if possible.

@olafurpg @tgodzik single option it is. the updated commits have been uploaded.

**Review by olafurpg — APPROVED** [review]
I think it would be nice to group all of the indentation-related settings under the same object. For example, we could rename `continuationIndent` into `indent` and introduce `indent.default` (exact name up for debate). 

Either way, LGTM.

**Review by tgodzik — APPROVED** [review]
LGTM! Thanks for taking care of this. Hopefully, we don't regret adding the option :sweat_smile:

**Comment by kitbellew** [comment]
> I think it would be nice to group all of the indentation-related settings under the same object. For example, we could rename `continuationIndent` into `indent` and introduce `indent.default` (exact name up for debate).

Done.

@olafurpg @tgodzik @poslegm please help me confirm that the first commit i added (removal of indent settings from the dynamic config) is not going to affect anything. (it's package-private, can't imagine why it would be visible outside the repo.)

File: scalafmt-core/shared/src/main/scala/org/scalafmt/config/ScalafmtConfig.scala [hunk 2]
```diff
@@ -105,7 +105,8 @@ case class ScalafmtConfig(
     comments: Comments = Comments(),
     optIn: OptIn = OptIn(),
     binPack: BinPack = BinPack(),
-    continuationIndent: ContinuationIndent = ContinuationIndent(),
+    @annotation.ExtraName("continuationIndent")
```
Comment (kitbellew) [inline scalafmt-core/shared/src/main/scala/org/scalafmt/config/ScalafmtConfig.scala:108]: old name is preserved

**Comment by olafurpg** [comment]
> please help me confirm that the first commit i added (removal of indent settings from the dynamic config) is not going to affect anything. 

The only public library API for Scalafmt is the Java interfaces in the `scalafmt-interfaces` module. Everything else can be considered private so feel free to remove/change methods/classes.

**Comment by olafurpg** [comment]
FWIW, `ScalafmtConfig` is a case class, which means that every addition of settings is a binary breaking change on the auto-generated `copy()` method. There shouldn't be any client that's calling those APIs directly outside of the Scalafmt repo.

**Pull request merged**